### ISSUE TITLE
Validate many JATS files as a batch and report on results

### ISSUE THREAD

**Issue by gnott** [elifesciences/elife-jats-2-dar-jats#3]
The JATS output this project produces can be validated against Texture. In future there may be a Command Line Interface (CLI) method of validating. Currently validation is limited to using Texture in the web browser to test validation.

We might be able to follow a procedure similar to on https://github.com/elifesciences/bot-lax-adaptor/blob/develop/generate-statistics.sh, using shell to provide the logic.

One aim is to discover which particular article XML files do not pass validation after they are converted. Reporting on how many articles pass validation, how many fail, and the list of articles that failed will be useful in tracking how much work remains in the converter development.

It would be ideal to also report on the validation errors for each of the articles that do no pass validation.

**Comment by gnott** [comment]
Today I started looking at https://github.com/substance/texture for some options for testing XML validation, in the simplest way possible, for these purposes here.

I was able to run or break the web browser portion by replacing the `data/kitchen-sink/manuscript.xml` file (as described in the Texture readme) prior to running `npm start`.

I also ran the project tests, and some of those fail when the `data/kitchen-sink/manuscript.xml` file is not a valid one.

With a tiny change I was able to run `https://github.com/substance/texture/blob/master/test/Persistence.test.js` only (I had to add in a line to import globals). The output is not very verbose, but it has a pass or fail.

Pass output looks like,

```
# Persistence: loading and saving article blank
ok 1 There should be no error.
ok 2 XML did not change
ok 3 The persisted file should be loaded and rendered without problems
# Persistence: loading and saving article kitchen-sink
ok 4 There should be no error.
ok 5 XML did not change
ok 6 The persisted file should be loaded and rendered without problems
# Persistence: saving and loading an article after inserting a figure
ok 7 The persisted file should be loaded and rendered without problems

1..7
# tests 7
# pass  7

# ok
```

Fail output looks like,

```
# Persistence: loading and saving article kitchen-sink
not ok 4 Uncaught error: Error: Error: JATS import failed
  ---
    operator: fail
  ...
# Persistence: saving and loading an article after inserting a figure
ok 5 The persisted file should be loaded and rendered without problems

1..5
# tests 5
# pass  4
# fail  1
```

So one possible method to test pass or fail of XML files in a slightly efficient way with the least amount of code could be:

1. Convert eLife JATS 2 DAR JATS
2. Copy it over to `data/kitchen-sink/manuscript.xml` (maybe in a docker shared volume or something?)
3. Run the Persistence test and capture the result

Do that thousands of times for every article XML we want to check.

**Comment by gnott** [comment]
An example of the command to run (after the globals line is added, we could put on a fork maybe),

```
node make test-nodejs && node --require esm test/Persistence.test.js
```

It means we wouldn't have to run the browser or capture web output in order to check validation pass or fail.

(~~can also probably drop the `node make test-nodejs` part if we `make` once and then run each test on that build~~ - need this so far, it seems to add the altered kitchen sink manuscript.xml to the right place to test properly).

**Comment by gnott** [comment]
Running the above takes about 4-5 seconds. I played with another little tweak we could cut it down to about 2 seconds by only doing `build:vfs` before we test each XML file.

**Comment by Melissa37** [comment]
@FAtherden-eLife WDYT?

**Comment by fred-atherden** [comment]
Sounds great! I think this is exactly what we need, since the Texture schema wasn't robust enough to cover all cases where the XML broke Texture. We can use this method to pick out those that fail, and investigate the cause.

**Comment by gnott** [comment]
Yesterday I put together quick example for converting XML with the Xquery (nothing to commit yet). Action to be done there is to sort out only the most recent versions for conversion because it took way too long.

Today I'm testing the validation again, with a few strange issues I didn't get in my first tests, but I'm getting some pass or fail now.

I am tentatively testing small batches of articles. I tried some early articles, and I don't think any of those passed. Next, I tried some around the range of your test article `42955`, which thankfully does pass the current test I'm using.

For some early results if you want to see if these match your results, and if the fail articles you can determine where the Xquery can be improved (the results for articles with id like `429xx`):

```
pass: elife-42914-v2.xml
pass: elife-42955-v2.xml
pass: elife-42976-v1.xml
pass: elife-42995-v3.xml
fail: elife-42908-v2.xml
fail: elife-42940-v1.xml
fail: elife-42974-v2.xml
fail: elife-42974-v3.xml
```

**Comment by fred-atherden** [comment]
Thanks @gnott this is really helpful. 42974 failed because it has a preprint type reference in the reference list. Apparently Texture can't handle that yet.

I've edited the XQuery to account for this in my latest commit. 

I can't reproduce the issue with 42940 and 42908, since both of these load for me locally (after having carried out the XQuery).

**Comment by gnott** [comment]
Thanks for checking those @FAtherden-eLife.

I converted `42974` again with the latest XQuery and it reported a pass from my drafted validation script.

`42940` I still get a `fail` from my script, but when I bring up the Texture web interface it is visible there. I will need to dig deeper now to figure out why the Persistence check fails but the web side of things works. The Persistence test may not be suitable for the purposes of bulk validation checking, or it is indicating something that the web side of things is not reporting on.

**Comment by gnott** [comment]
I got a hint about why the persistence tests for `42940` do not pass, when running the test directly it provides an error message like this:

```
      { _errors: [ { errors: [ { el: [Object], msg: '<ext-link> is not valid in <italic>\nSchema: (TEXT|bold|fixed-case|italic|monospace|overline|overline-start|overline-end|roman|sans-serif|sc|strike|underline|underline-start|underline-end|ruby|sub|sup)*' }, { el: [Object], msg: '<ext-link> is not valid in <italic>\nSchema: (TEXT|bold|fixed-case|italic|monospace|overline|overline-start|overline-end|roman|sans-serif|sc|strike|underline|underline-start|underline-end|ruby|sub|sup)*' }, { el: [Object], msg: '<ext-link> is not valid in <italic>\nSchema: (TEXT|bold|fixed-case|italic|monospace|overline|overline-start|overline-end|roman|sans-serif|sc|strike|underline|underline-start|underline-end|ruby|sub|sup)*' }, { el: [Object], msg: '<ext-link> is not valid in <italic>\nSchema: (TEXT|bold|fixed-case|italic|monospace|overline|overline-start|overline-end|roman|sans-serif|sc|strike|underline|underline-start|underline-end|ruby|sub|sup)*' }, { el: [Object], msg: '<ext-link> is not valid in <italic>\nSchema: (TEXT|bold|fixed-case|italic|monospace|overline|overline-start|overline-end|roman|sans-serif|sc|strike|underline|underline-start|underline-end|ruby|sub|sup)*' } ], name: 'validate-texture-article' } ] }
```

The hint being to look for `<ext-link>` tags. I found some in the table cells of the Materials and Methods table in article `42940`, for RRID links. When I removed those `<ext-link>` tags then the validation test I have returns a `pass`.

We could break tables out into a separate issue if you feel like that is a good idea @FAtherden-eLife. At a minimum, maybe you could please consider to write some XQuery to remove `<ext-link>` tags from tables?

Also to please note, although the `<italic>` and `<sup>` tags in the table do not confuse my validation script, I do not see those styles retained when viewing this Materials and Methods table in Texture. I don't know the level of support for inline styles in tables in Texture. The GUI seems to allow setting them, but the styles may not be read in from the XML file when loaded (not sure).

**Comment by gnott** [comment]
Just a note, I merged the latest `master` branch of the `texture` project over to the elifesciences fork so the next set of validation results will be based on the latest Texture code available as of now.

**Comment by gnott** [comment]
I think there was something wrong with the latest version of Texture, when I tried it again, so I changed it back to the earlier working version until I can spend more time figuring out why the latest didn't work.

I've committed some of my early code to a `validate` branch for now https://github.com/elifesciences/elife-jats-2-dar-jats/tree/validate. It has a few commits since I forgot to add the `output` folder, and more importantly, I forgot to commit my changes to the Xquery file.

I think it might be possible for someone else to run it. There are many todo items to make it possible run as a batch in an automated test, but at least it is something to start from. 

It might be possible to run it locally as I am doing for investigations and requirements gathering.

Running it automatically would probably be best to run `zorba` and `texture` / `node` in docker containers. It might be slow too without some parallel running of tests.
